controllers/subnamespace_controller.go

```diff
@@ -34,18 +36,105 @@ type SubnamespaceReconciler struct {
+	if err := r.Get(ctx, req.NamespacedName, &subspace); err != nil {
+		log.Error(err, "Could not find Subspace")
+		return ctrl.Result{}, err
+	}
+	// Getting subspace's namespace - described as ownerNamespace
+	if err := r.Get(ctx, client.ObjectKey{Namespace: "", Name: subspace.GetNamespace()}, &ownerNamespace); err != nil {
+		log.Error(err, "Could not find owner Namespace ")
+		return ctrl.Result{}, err
+	}
+
+	// Hierarchy 'breadcrumb' name, built from ownerNamespace and subspace.
+	childNamespaceName = ownerNamespace.Name + "-" + subspace.Name
 
-func (r *SubnamespaceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("subnamespace", req.NamespacedName)
+	// Subspace phase None flow
+	if subspace.Status.Phase == danav1alpha1.None {
 
-	// your logic here
+		if _, err := ctrl.CreateOrUpdate(ctx, r.Client, &subspace, func() error {
+			err := InitializeSubspace(ownerNamespace, subspace, childNamespaceName)
+			return err
+		}); err != nil {
+			log.Error(err, "Could not update Subspace Phase from 'None' to 'Missing'")
+		}
+	}
 
+	// Subspace phase Missing flow
+	if subspace.Status.Phase == danav1alpha1.Missing {
+		childNamespace = GetNewChildNamespace(ownerNamespace.Name, subspace.Name, childNamespaceName)
+		if err := r.Get(ctx, client.ObjectKey{Namespace: "", Name: childNamespaceName}, &childNamespace); err != nil {
+			if errors.IsNotFound(err) {
+				// Child Namespace is not found, therefore we are creating one
+				if err := r.Create(ctx, &childNamespace); err != nil {
+					log.Error(err, "Could not create Namespace ")
+					return ctrl.Result{}, err
+				}
+			}
+		} else {
+			// Child Namespace is exist, so we are changing the Subspace phase to created
+			if _, err := ctrl.CreateOrUpdate(ctx, r.Client, &subspace, func() error {
+				subspace.Status.Phase = danav1alpha1.Created
+				return nil
+			}); err != nil {
+				log.Error(err, "Could not update Subspace Phase from 'Missing' to 'Created'")
+			}
+		}
+
+	}
+
+	// Subspace phase Created flow
+	if subspace.Status.Phase == danav1alpha1.Created {
+		if err := r.Get(ctx, client.ObjectKey{Namespace: "", Name: childNamespaceName}, &childNamespace); err != nil {
+			if errors.IsNotFound(err) {
+				if _, err := ctrl.CreateOrUpdate(ctx, r.Client, &subspace, func() error {
+					subspace.Status.Phase = danav1alpha1.Missing
+					return nil
+				}); err != nil {
+					log.Error(err, "Could not update Subspace Phase from 'Created' to 'Missing'")
+				}
+			}
+		}
+	}
 	return ctrl.Result{}, nil
 }
+func GetNewChildNamespace(ownerNamespaceName string, subspaceName string, childNamespaceName string) v1.Namespace {
+	return v1.Namespace{
+		ObjectMeta: v1api.ObjectMeta{
+			Name: childNamespaceName,
+			Labels: map[string]string{
+				danav1alpha1.Hns:    "true",
+				danav1alpha1.Parent: ownerNamespaceName,
+			},
+			Annotations: map[string]string{
+				danav1alpha1.SnsPointer: subspaceName,
+			},
+		},
+	}
+}
```

Comment:
I like the idea of isolating the creating of an object but should we move it to a different file?

Reply:
That could be nice.
In which file do you think this could fit in ?